app/models/token.rb

```diff
@@ -0,0 +1,73 @@
+# frozen_string_literal: true
+
+class Token
+  attr_reader :token_string
+
+  def initialize(token_string)
+    @token_string = token_string
+    validate_token
+  end
+
+  def to_s
+    @token_string
+  end
+
+  def payload
+    @payload ||= if @token_string
+                   pubkey = expected_key
+                   return if pubkey.blank?
+
+                   JWT.decode(@token_string, pubkey, true, algorithm: 'RS256')[0]
+                 end
+  rescue JWT::ExpiredSignature
+    raise error_klass('Validation error: token has expired')
+  rescue JWT::DecodeError
+    raise error_klass('Validation error: token could not be decoded')
```

Comment:
There's a lot of different kinds of `JWT::DecodeError` that could be raised here (such as https://github.com/jwt/ruby-jwt/blob/v2.1.0/lib/jwt/verify.rb#L32-L82). It looks to me like the messages from these errors are generally suitable for raising to the user, so could we do that?

If that makes you queasy, could we at least log the more specific error details here?